text/3334-niche.md

```diff
@@ -0,0 +1,480 @@
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Provide a stable attribute to define "niche" values of a type. The type cannot
+store these values, allowing the compiler to use them to optimize the
+representation of containing types such as `Option<Type>`.
+
+# Motivation
+[motivation]: #motivation
+
+Rust makes extensive use of types like `Option`, and many programs benefit from
+the efficient storage of such types. Many programs also interface with others
+via FFI, via interfaces that provide data and a sentinel value (such as for
+errors or missing data) within the same bits.
+
+The Rust compiler already provides support for this via "niche" optimizations,
+and various types providing guarantees of such optimizations, including
+references, `bool`, `char`, and the `NonZero` family of types. However, Rust
+does not provide any stable means of defining new types with niches, reserving
+this mechanism for the standard library. This puts pressure on the standard
+library to provide additional families of types with niches, while preventing
+the broader crate ecosystem from experimenting with such types.
+
+Past efforts to define a stable niche mechanism stalled out due to scope creep:
+alignment niches, null-page niches, multiple niches, structures with multiple
+fields, and many other valid but challenging ideas (documented in the "Future
+possibilities" section). This RFC defines a *simple* mechanism for defining one
+common type of niche, while leaving room for future extension.
+
+Defining a niche mechanism allows libraries to build arbitrary types containing
+niches, and simplifies handling of space-efficient data structures in Rust
+without manual bit-twiddling.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a struct containing exactly one field, you can attach a `niche`
+attribute to the struct to declare a specific value or range of values for that
+field as invalid. This promises the compiler that you will never store those
+values in that field, which allows the compiler to use those in-memory
+representations for different purposes, such as the representation of `None` in
+a containing `Option`.
+
+```rust
+use std::mem::size_of;
+
+#[niche(value = 42)]
+struct MeaninglessNumber(u64);
+
+assert_eq!(size_of::<MeaninglessNumber>(), 8);
+assert_eq!(size_of::<Option<MeaninglessNumber>>(), 8);
+
+#[niche(range = 2..)]
+struct Bit(u8);
+
+assert_eq!(size_of::<Bit>(), 1);
+assert_eq!(size_of::<Option<Option<Option<Bit>>>>(), 1);
+```
+
+Constructing a struct with a niche value, or writing to the field of such a
+struct, or obtaining a mutable reference to such a field, requires `unsafe`
+code. Causing a type with a niche to contain one of its niche values (whether
+by construction, writing, or transmuting) results in undefined behavior.
+
+The field type must be a built-in integer or floating-point type, a `char`, or
```

Comment:
It is worth clarifying that `char` already has a restricted validity range, and this mechanism can only be used to restrict it further -- it can never be used to *relax* a validity range of a field type.